### DISCUSSION TITLE
Batch Image Merging

### DISCUSSION THREAD

**Discussion by uppearl** [ImageMagick/ImageMagick#2711] [Help]
I would like to merge a batch of images onto one which is supposed to be the background for the rest.

For individual merging, I used the following and it worked.

magick convert image1.png image2.png -gravity center -geometry -50+100 -composite result.png

I tried to batch process the images in the folder using the following. The script ran but it didn't work as intended. image1.png which is the background, is in the same folder.

magick mogrify image1.png *.png -gravity center -geometry -50+100 -composite

I'm using IM v7.0 on Windows 10.

Thanks in advance.

**Comment by snibgo** [comment]
One command works, and the other doesn't, but the two commands are the same???

> ... it didn't work as intended ...

From that description, diagnosis is difficult. I suggest you post sample input images.

Please also say what version of IM, on what platform.

**Reply by uppearl** [reply]
My mistake. I was in a rush so I wrote the wrong second command. Was using my phone and tried to copy and paste the second command but somehow it didn't copy properly.

The second command is:

magick mogrify image1.png *.png -gravity center -geometry -50+100 -composite

I'm using IM v7.0 on Windows 10.

**Comment by snibgo** [comment]
Your "magick mogrify" syntax is wrong. But I don't think you can do what you want with mogrify. I suggest a simple "magick" command in a shell "for" loop. Write the outputs to a different directory. Windows BAT syntax:
```
for %%F in (*.png) do magick %%F image2.png -gravity center -geometry -50+100 -composite outdir\%%F
```
If running this directly in the command line, don't double the percent signs %.

If your inputs are small enough to all fit in memory at the same time, you could instead do something like this:
```
magick *.png NULL: image2.png -gravity center -geometry -50+100 -layers composite outdir\compout_%%03d.png
```

**Reply by uppearl** [reply]
I'm sorry I'm a programming noob.

I tried the first command with a single %.

for %F in (*.png) do magick %F image1.png -gravity center -geometry -50+100 -composite outdir\%F

I changed image2.png in your example to image1.png because I wanted image1.png to be the background. Yes, I wrote it exactly like that and now I know I shouldn't because I didn't specify the output folder but I've no idea how to do that even.

Anyway, the command ran and this is word-for-word one of the error for one of the images (I only modified my PC name):

C:\Users\myname\Desktop\Test>magick welder rawrsome.png blank_rawrsome_mug_mockup_2.PNG -gravity center -geometry -535.5+1069.5 -composite outdir\welder rawrsome.png
magick: unable to open image 'welder': No such file or directory @ error/blob.c/OpenBlob/3537.
magick: no decode delegate for this image format `' @ error/constitute.c/ReadImage/562.

"welder rawrsome.png" is a picture of a mug which is supposed to go on "blank_rawrsome_mug_mockup_2.PNG" which is a background picture I want the mug to go on.

Let's say I want the output to go on TestOutput folder on my Desktop, what should I write?

Also how do I determine if my inputs are small enough to fit in memory? All the images are 30 MB. So if my RAM is more than that, I can use the second command you wrote?

**Reply by t246246** [reply]
Your filename include space, that causes error.  Probably you need "%F" instead o %F.

**Reply by uppearl** [reply]
Changing %F to "%F" gives the following error message.

"%F" was unexpected at this time.

**Reply by uppearl** [reply]
Ok so I figured out how to get the images to export to another folder (changing outdir to the folder's address). I also removed the spaces in my file names. Below is my command:

for %F in (*.PNG) do magick %F blank_rawrsome_mug_mockup_2.PNG -gravity center -geometry -535.5+1069.5 -composite \Users\myname\Desktop\Test2\%F

The problem is the background image (blank_rawrsome_mug_mockup_2.PNG) seems to be layered over the mug image. If I input the name of a mug image in the command, then with the exception of one image, i.e. the background image, everything else simply outputs as the mug image itself. The background image gets the first mug image layered on it.

**Reply by t246246** [reply]
Quote %f only in do block.  And *.png will be expanded all the file in current folder that ended with .png.
You can exclude it using "if" syntax.

your command **in .bat file** will be like this, replace ... part in your own command

```
for %%f in (*.png) do (
  if not "%%f" == "background.png" (
    magick "%%f" ... outdir\"%%f"
  )
)
```

**Reply by uppearl** [reply]
I finally figured it out! I just had to reorder the %F behind "blank_rawrsome_mug_mockup_2.PNG" and it worked. Thank you so much!!
